src/charm.py

```diff
@@ -3,61 +3,77 @@
 import sys
 sys.path.append('lib')
 
-from ops.charm import CharmBase
-from ops.framework import StoredState
+from ops.charm import CharmBase, CharmEvents
+from ops.framework import EventBase, EventSource, StoredState
 from ops.main import main
 from ops.model import (
     ActiveStatus,
-    BlockedStatus,
-    WaitingStatus,
     MaintenanceStatus,
+    WaitingStatus,
 )
 
-from interfaces import (
-    HTTPInterfaceProvides,
-    MySQLInterfaceRequires,
-)
-from resources import OCIImageResource
+from interface_http import HTTPServer
+from interface_mysql import MySQLServer, MySQLServerError
+from oci_image import OCIImageResource, FetchError
+
+
+class ConfigurePodEvent(EventBase):
+    pass
+
+
+class ServeWebsiteEvent(EventBase):
+    pass
+
+
+class GitLabK8sCharmEvents(CharmEvents):
+    configure_pod = EventSource(ConfigurePodEvent)
+    serve_website = EventSource(ServeWebsiteEvent)
+
+
+class NotLeaderError(Exception):
+    def __init__(self):
+        super().__init__('not leader')
+        self.status = WaitingStatus('Not leader')
 
 
 class GitLabK8sCharm(CharmBase):
     state = StoredState()
+    on = GitLabK8sCharmEvents()
 
     def __init__(self, framework, key):
         super().__init__(framework, key)
 
-        for event in (self.on.start,
-                      self.on.upgrade_charm,
-                      self.on.config_changed,
-                      self.on.mysql_relation_changed):
-            self.framework.observe(event, self.on_start)
-        self.framework.observe(self.on.website_relation_joined, self)
-
-        self.website = HTTPInterfaceProvides(self, 'website')
-        self.mysql = MySQLInterfaceRequires(self, 'mysql')
+        self.website = HTTPServer(self, 'website')
+        self.mysql = MySQLServer(self, 'mysql')
         self.gitlab_image = OCIImageResource(self, 'gitlab_image')
 
+        self.state.is_started = getattr(self.state, 'is_started', False)
+        self.state.has_website = getattr(self.state, 'has_website', False)
```

Comment:
This will flag the state as changed/dirty although it hasn't. We need a better pattern for people to follow for "I need to initialize that state".

Reply:
Previously we discussed doing something like this in `on_install` or `on_start`, however, we could do something better in the framework itself.

Utilizing a constructor for `StoredState` comes to mind.

Reply:
I'd really like a better way of specifying initial / default state values in a clean way.  The `getattr()` pattern won't actually overwrite the value if it's been changed, but it's not clean at all.  Doing it in an `on_install` hook feels messy as well and prone to being forgotten or put into the wrong spot.

Reply:
Moved to an `on_start` handler (`init_state`), but I'd really like to discuss and explore further the ideas down in the thread from John's comment.